### PR TITLE
S3 streamer

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**

### CODE REVIEW

File: s3-streamer [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+            print(response.status, filename)
+
+    def delete(self, filenames):
+        # to do: multi-object delete API
+        for filename in filenames:
+            with s3.urlopen(self.url(filename), method='DELETE') as response:
+                print(response)
+
+
+class Index(Destination):
```
Comment (martinpitt) [inline s3-streamer:94]: Constantly maintaining an index.html, from dozens of parallel bots feels error prone and like unnecessary churn. We have both the S3 CLI and our test-results.db for querying available log URLs. I'd rather drop this.

File: s3-streamer [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+        for entry in os.scandir(self.path):
+            if not self.destination.has(entry.name) and entry.is_file(follow_symlinks=False):
+                with open(entry.path, 'rb') as file:
+                    data = file.read()
+                self.destination.write(entry.name, data)
+
+
+class ChunkedUploader:
+    SIZE_LIMIT = 1000000  # 1MB
+    TIME_LIMIT = 10       # 10s
```
Comment (martinpitt) [inline s3-streamer:145]: This was probably fine for testing, but for production one minute feels enough. Our logs are not that huge, so of these two it'll be the time limit which has the practical effect.

File: s3-streamer [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+    def __init__(self, index, filename, encoding=None):
+        self.input_decoder = codecs.getincrementaldecoder(encoding or locale.getpreferredencoding())()
```
Comment (martinpitt) [inline s3-streamer:148]: I'd rather not depend on the locale here. This has bitten us all too often in the past, with bots running under `C` locale and regurgitating UTF-8 input.

As I said, I'd rather drop that whole decode → encode dance completely and just handle byte streams, with breaking on the last `b\n` of a read block (as sending half a line isn't that useful anyway). This avoids potential locale crashes, hangs with internal buffers of the encoders, and saves CPU cycles on top of that. You seemed very eager to do the encode/decode, but let's at least do a fixed `utf-8` here.

**Review by martinpitt — COMMENTED** [review]
Cheers! Some initial questions/comments.

File: s3-streamer (hunk 1 above)
Comment (allisonkarlitskaya) [reply s3-streamer:100]: Each bot generates a unique subdirectory name and writes to it.  We never try to maintain an index.html for the toplevel.

All the same, I may indeed drop this code...

File: s3-streamer (hunk 3 above)
Comment (allisonkarlitskaya) [reply s3-streamer:159]: Let's chat about this.

My priorities here are:
 - it needs to be impossible to send a partial multi-byte character
 - i also want to never ever send non-utf8 data to the client.  that means we need to detect invalid sequences and transform them, anyway.
 - we should see all of the available output in the event of a hang.

The locale is going to be utf-8 in any reasonable universe.  I use the user's preferred locale out of a desire to be strictly correct.  I can turn that into an assert, if you like.

File: s3-streamer (hunk 2 above)
Comment (allisonkarlitskaya) [reply s3-streamer:156]: Keep in mind that this doesn't mean that we send a block every 10 seconds.  It means that after we see some data, we wait for more data for 10 seconds before sending the accumulated data as a block.

I thought we could start low and increase it if we find that we're sending too much data...

File: s3-streamer (hunk 1 above)
Comment (allisonkarlitskaya) [reply s3-streamer:100]: I want to come up with a replacement before I drop this, though.  In particular, I want to send a static `index.html` which queries the REST API for the file list and generates an index on the client side.  That shouldn't be hard, but it's not on the critical path to landing this, so I'd rather do it later.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks, this is exciting! Nothing that blocks first experiments in a manual container, but let's resolve the discussion threads before landing.

**Review by martinpitt — APPROVED** [review]

**Pull request merged**